cmd/main.go

```diff
@@ -19,115 +19,27 @@ package main
 import (
 	"log"
 
-	"github.com/spf13/cobra"
-
-	"sigs.k8s.io/kubebuilder/cmd/internal"
 	"sigs.k8s.io/kubebuilder/cmd/version"
-	"sigs.k8s.io/kubebuilder/internal/config"
-	"sigs.k8s.io/kubebuilder/pkg/scaffold"
+	"sigs.k8s.io/kubebuilder/pkg/cli"
+	pluginv1 "sigs.k8s.io/kubebuilder/pkg/plugin/v1"
+	pluginv2 "sigs.k8s.io/kubebuilder/pkg/plugin/v2"
 )
 
-// commandOptions represent the types used to implement the different commands
-type commandOptions interface {
-	// bindFlags binds the command flags to the fields in the options struct
-	bindFlags(command *cobra.Command)
-
-	// The following steps define a generic logic to follow when developing new commands. Some steps may be no-ops.
-	// - Step 1: load the config failing if expected but not found or if not expected but found
-	loadConfig() (*config.Config, error)
-	// - Step 2: verify that the command can be run (e.g., go version, project version, arguments, ...)
-	validate(*config.Config) error
-	// - Step 3: create the Scaffolder instance
-	scaffolder(*config.Config) (scaffold.Scaffolder, error)
-	// - Step 4: call the Scaffold method of the Scaffolder instance
-	// Doesn't need any method
-	// - Step 5: finish the command execution
-	postScaffold(*config.Config) error
-}
-
-// run executes a command
-func run(options commandOptions) error {
-	// Step 1: load config
-	projectConfig, err := options.loadConfig()
-	if err != nil {
-		return err
-	}
-
-	// Step 2: validate
-	if err := options.validate(projectConfig); err != nil {
-		return err
-	}
-
-	// Step 3: create scaffolder
-	scaffolder, err := options.scaffolder(projectConfig)
+func main() {
+	c, err := cli.New(
+		cli.WithPlugins(
+			&pluginv1.Plugin{},
+			&pluginv2.Plugin{},
+		),
+		cli.WithExtraCommands(
+			newEditCmd(),
```

Comment:
There are two options here to avoid naming conflicts:
1. Create an `EditPlugin(Getter)` interface set.
1. Have `edit` implemented as a standalone subcommand added by `WithExtraCommands` in `cmd/`.

If we add `edit` directly in `buildRootCmd` we prevent downstream users from implementing their own `edit` command:

```go
func (c *cli) initialize() error {
	...
	// rootCmd will already contain any plugin and directly added
	// command names, so any conflicting names in c.extraCommands
	// will result in an error.
	rootCmd := c.buildRootCmd()
	for _, cmd := range c.extraCommands {
		for _, subCmd := range rootCmd.Commands() {
			if cmd.Name() == subCmd.Name() {
				return fmt.Errorf("command %q already exists", cmd.Name())
			}
		}
		rootCmd.AddCommand(cmd)
	}
	...
}
```

Generally any command that we don't want to reserve for kubebuilder alone (so pretty much every command) should not be added directly _without a plugin interface_ to `buildRootCmd`. Currently `update` does this, and should be refactored using one of the 2 methods above.

Reply:
I think update is something too specific (and deprecated) so let it as is. Edit, however, should be open for other plugins to use. Thats my opinion.

Reply:
@DirectXMan12 @mengqiy thoughts on `update`?

Reply:
I saw your last commit and don't know if I was clear enough with my last comment. I would leave update as a extra command but I would provide an interface to edit as it whould allow plugins have their own edit subcommands. However, I would leave all commands in either `cli` os `plugin` packages, not in `cmd`.

Reply:
I moved update to `cmd` so we can merge things as-is without inciting unnecessary conflict while we decide how to handle these commands. It would be best to do so in a follow-up IMO.